### PR TITLE
Intel RDT plugin: add support for groups of PIDs.

### PR DESCRIPTION
**Pull request by wandralx**
ChangeLog: intel_rdt: Add support for groups of PIDs monitoring

intel_rdt plugin takes processes’ names as single items or as groups.
Plugin will reconfigure pqos library when given process is terminated or created.

For each process or group of processes following statistics are collected:
- LLC occupancy,
- IPC,
- MBL and MBR.

### CODE REVIEW

**Comment by wandralx** [comment]
@rpv-tomsk any thoughts/comments? when can I expect PR to be accepted? Thank you!

**Comment by wandralx** [comment]
@rubenk any thoughts/comments? when can I expect PR to be accepted? Thank you!

**Comment by rpv-tomsk** [comment]
Can I check this plugin on Core2Duo CPU ?

```
# cat /proc/cpuinfo
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 23
model name      : Intel(R) Core(TM)2 Duo CPU     E8200  @ 2.66GHz
stepping        : 6
cpu MHz         : 2662.464
cache size      : 6144 KB
fdiv_bug        : no
hlt_bug         : no
f00f_bug        : no
coma_bug        : no
fpu             : yes
fpu_exception   : yes
cpuid level     : 10
wp              : yes
flags           : fpu de tsc msr pae cx8 sep cmov pat clflush mmx fxsr sse sse2 ss ht nx constant_tsc aperfmperf pni ssse3 sse4_1 hypervisor
bogomips        : 5324.92
clflush size    : 64
cache_alignment : 64
address sizes   : 36 bits physical, 48 bits virtual
power management:
```

**Comment by wandralx** [comment]
@rpv-tomsk thank you for your interest!

Unfortunately no, to be able to check this plugin, you need one of Intel(R) Xeon(R) processors.

For detailed info about required hardware and kernel versions please see README file:
https://github.com/intel/intel-cmt-cat

For info about /proc/cpuinfo flag bits please see:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/x86/intel_rdt_ui.txt

File: src/intel_rdt.c [hunk 1]
```diff
@@ -85,39 +154,726 @@ static void rdt_dump_cgroups(void) {
...
+  /* Get all available events on this platform */
+  for (unsigned i = 0; i < rdt->cap_mon->u.mon->num_events; i++)
+    events |= rdt->cap_mon->u.mon->events[i].type;
+
+  events &= ~(PQOS_PERF_EVENT_LLC_MISS);
+
+  DEBUG(RDT_PLUGIN ": Available events to monitor: %#x", events);
+
+  rdt->num_ngroups = n;
+  for (int i = 0; i < n; i++) {
```
[the remark this answers is not in the document]
Comment (aleksinx) [reply src/intel_rdt.c]: This helper function is called by rdt_config only

**Comment by sunkuranganath** [comment]
Thanks @rpv-tomsk for the review. Working on these internally. Will have few updates very soon.

**Comment by rpv-tomsk** [comment]
Great news!

File: src/intel_rdt.c [hunk 2]
```diff
@@ -85,39 +154,726 @@ static void rdt_dump_cgroups(void) {
...
+  }
+
+#if COLLECT_DEBUG
+  rdt_dump_pids_data();
 #endif /* COLLECT_DEBUG */
 
+groups_refresh:
+  for (size_t i = 0; i < g_rdt->num_ngroups; i++) {
+    int refresh_result =
+        rdt_refresh_ngroup(&(g_rdt->ngroups[i]), g_rdt->pngroups[i]);
```
[the remark this answers is not in the document]
Comment (aleksinx) [reply src/intel_rdt.c]: proc_pids_update moved before the for loop. Now list of pids is fetched only once.

File: src/intel_rdt.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (rpv-tomsk) [reply src/intel_rdt.c]: >This implementation does not take into account that **one option can be used in config several times**.

[the remark this answers is not in the document]
Comment (aleksinx) [reply src/intel_rdt.c]: > > This implementation does not take into account that **one option can be used in config several times**.

Yes, currently we support only single Processes or Cores option. We plan to extend that in the future.

**Comment by kwiatrox** [comment]
Hi @rpv-tomsk,  Currently we support only single use of Processes or Cores option. This is the case also for some other plugins. It's a good idea, but maybe we should improve it in the future, not in this PR.
We just would like to include it in 5.9 release.

**Comment by sunkuranganath** [comment]
I would go with @kwiatrox suggestion. Could we have this accepted for 5.9 release and have the updates done just after as a new PR?

**Comment by rpv-tomsk** [comment]
>Hi @rpv-tomsk, Currently we support only single use of Processes or Cores option. 
>This is the case also for some other plugins. It's a good idea, but 
>**maybe** we should improve it in the future, not in this PR.
>We just would like to include it in 5.9 release.


I just want to include *only* qualitative code. I think there is no much code changes needed to allow multiple use of `Processes` option, or to deny such explicitly. This reduces configuration errors.
From my POV "maybe" and "in the future" are not allowed for such small changes.

**Comment by sunkuranganath** [comment]
Thanks @rpv-tomsk. We see 5.9 release hasnt happened yet and so we are working on the updates as indicated.

**Comment by rpv-tomsk** [comment]
Hi, Team.

I will not merge none of PR untill it will be possible for me to merge my PR too.

Thanks for your attention.

**Comment by mrunge** [comment]
Thank you all for your contribution.

**Pull request merged**